File: dsa_algo/patterns/fast_slow_pointers/1_linked_list_cycle.py
class node:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next
def has_cycle(head):
    slow, fast = head, head
    while fast is not None and fast.next is not None:
        fast = fast.next.next
        slow = slow.next
        if slow == fast:
            return True
    return False

def main():
    head = node(1)
    head.next = node(2)
    head.next.next = node(3)
    head.next.next.next = node(4)
    head.next.next.next.next = node(5)
    head.next.next.next.next.next = node(6)
    print("List has cycle: " + str(has_cycle(head)))

    head.next.next.next.next.next.next = head.next.next
    print("List has cycle: " + str(has_cycle(head)))

    head.next.next.next.next.next.next = head.next.next.next
    print("List has cycle: " + str(has_cycle(head)))

main()

File: dsa_algo/patterns/fast_slow_pointers/1_linked_list_cycle.ts
class node {
    value: number;
    next: node | null;
    constructor(value: number, next: node | null = null)
    {
        this.value = value;
        this.next = next;
    }
}

const has_cycle = (head: node | null) =>{
    let slow = head;
    let fast = head;
    while(fast !== null && fast.next !== null)
    {
        fast = fast.next.next;
        if(slow)
            slow = slow.next;
        if(slow === fast)
            return true;
    }
    return false;
}

const head = new node(1);
head.next = new node(2);
head.next.next = new node(3);
head.next.next.next = new node(4);
head.next.next.next.next = new node(5);
head.next.next.next.next.next = new node(6);
console.log(`list has cycle: ${has_cycle(head)}`);

head.next.next.next.next.next.next = head.next.next;
console.log(`list has cycle: ${has_cycle(head)}`);

head.next.next.next.next.next.next = head.next.next.next;
console.log(`list has cycle: ${has_cycle(head)}`);
